kythe/cxx/indexer/textproto/plugin.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2020 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef KYTHE_CXX_INDEXER_TEXTPROTO_PLUGIN_H_
+#define KYTHE_CXX_INDEXER_TEXTPROTO_PLUGIN_H_
+
+#include <cstdio>
+#include <string>
+
+#include "absl/status/status.h"
+#include "kythe/cxx/common/file_vname_generator.h"
+#include "kythe/cxx/common/indexing/KytheGraphRecorder.h"
+#include "kythe/cxx/common/status_or.h"
+#include "kythe/cxx/indexer/proto/vname_util.h"
+#include "kythe/proto/analysis.pb.h"
+#include "re2/re2.h"
+
+namespace kythe {
+namespace lang_textproto {
+
+// The plugin's interface to the indexer.
+class PluginApi {
+ public:
+  virtual proto::VName CreateAndAddAnchorNode(const proto::VName& file,
+                                              int begin, int end) = 0;
+
+  virtual proto::VName CreateAndAddAnchorNode(const proto::VName& file_vname,
+                                              re2::StringPiece sp) = 0;
+
+  virtual KytheGraphRecorder* recorder() = 0;
+
+  virtual void EmitDiagnostic(const proto::VName& file_vname,
+                              absl::string_view signature,
+                              absl::string_view msg) = 0;
+
+  virtual absl::optional<proto::VName> VNameForRelPath(
+      absl::string_view simplified_path) const = 0;
+
+  // Create a VName for a proto descriptor.
+  template <typename SomeDescriptor>
+  StatusOr<proto::VName> VNameForDescriptor(const SomeDescriptor* descriptor) {
```

Comment:
This should probably be a namespace scoped function rather than a member function.

Reply:
I would generally agree, but this relies on the `VNameForRelPath()` member function, which uses the file substitution cache for mapping paths to proto files.

Reply:
It still works fine as a non-member function, it just has to accept an instance of `PluginApi`. It could/should live alongside kythe::lang_proto::VNameForDescriptor. The only different between the two is that this one allows for failure, which should arguably be the default.

Reply:
Related: https://github.com/kythe/kythe/pull/4627. The *proto* indexer's vname lookup functions return an empty vname if given an invalid path. I could change those  to also return `StatusOr` alongside `kythe::lang_proto::VNameForDescriptor()`.

I think ideally in both cases (proto and textproto), the initial parse would fail if it couldn't find a dependent file and the indexing code wouldn't run. If the indexer isn't able to find a file vname, it indicates a bug in the indexer, rather than invalid inputs. Maybe we could just log a "this is a bug" error and return an empty vname?

Reply:
That seems reasonable to me.

Reply:
I removed the StatusOr and added a log statement in case something does request a vname for an invalid path. I also removed `VNameForDescriptor()` from the plugin api.